maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/report/TestSetRunListener.java

```diff
@@ -312,7 +312,7 @@ private void addTestMethodStats()
         for ( WrappedReportEntry reportEntry : detailsForThis.getReportEntries() )
         {
             TestMethodStats methodStats =
-                new TestMethodStats( reportEntry.getClassMethodName(), reportEntry.getReportEntryType(),
+                new TestMethodStats( reportEntry.getReportClassMethodName(), reportEntry.getReportEntryType(),
```

Comment:
@chalmagr I do not think this is right because source/name is always extracted from JUnit4 Description which encodes both strings as follows `name(source)` and JUnit ensures that the descrption is unique. Typically, junit encodes `name` as `<method>[<id>]`. If this is wrong, then the problem is in provider and not in this class TSRL.

Reply:
I may need to dive a bit more and verify, but seems that even though JUnit is providing the source as you describe we are removing the (source) piece and splitting it between class and method

org.apache.maven.surefire.common.junit4.JUnit4RunListener#testFailure
```java
    ...
            ClassMethod classMethod = toClassMethod( failure.getDescription() );
            long testRunId = classMethodIndexer.indexClassMethod( classMethod.getClazz(), classMethod.getMethod() );
            ReportEntry report = withException( runMode, testRunId, classMethod.getClazz(), null,
                classMethod.getMethod(), null, stackTrace );
    ...
```

org.apache.maven.surefire.common.junit4.JUnit4ProviderUtil#toClassMethod
```java
public static ClassMethod toClassMethod( Description description )
    {
        String clazz = extractClassName( description.getDisplayName() );
    
    ...
        String method = extractMethodName( description.getDisplayName() );
        return new ClassMethod( clazz, method );
    }
```

org.apache.maven.surefire.api.util.internal.TestClassMethodNameUtils#extractClassName

```java
    public static String extractClassName( String displayName )
    {
        String clazz = displayName;
        if ( displayName.endsWith( ")" ) )
        {
            int paren = displayName.lastIndexOf( '(' );
            if ( paren != -1 )
            {
                clazz = displayName.substring( paren + 1, displayName.length() - 1 );
            }
        }
        return clazz;
    }
```

Reply:
Make sure you understand b2cce57a5e1dbd9a3a9e79c2eadd866e289e6f7f and especially e8f65b99aaaa7c04e1ad3770309cd626d3fe5ab0. I believe the fix for this issue should be in RunListenerAdapter. The logic there is hard to reason about without comments describing intent but ultimately I think you'll want methodDesc to be set to the legacy reporting name (given by junit in the testidentifier directly) for these junit4 parameterized tests, not methodSource.getMethodName() as it is today. It's using the method name because displayName and legacyReportingName are found to be equal (both e.g. methodName[0]).

When running junit5 parameterized tests, this does not repro because junit creates the testidentifier setting the legacy reporting name to methodName(ParameterDataType1, ..)[testInvocationNum] (e.g. "methodName(Long)[0]") and the display name if not overridden (@ParameterizedTest(name="blah")) to [testInvocationNum] argList e.g. "[0] 100" if the given param is 100L and it's the first parameterized case. Since the code finds them unequal it sets methodDesc to the legacyReportingName. The basic issue here is that the handling for parameterized tests makes an implicit assumption that they will always have a different legacy reporting name from display name which isn't the case for junit 4 parameterized tests without the @Parameters name overridden.

Reply:
Right... Was looking into my previous thought and was able to confirm that it was indeed ok and wasn't really affecting it as it depends on the provider... eventually ended up in this RunListenerAdapter class as well, and the logic there is quite overwhelming and a bit scary :) I have made changes to that class only now (without all the other changes) and it does seem to be ok, as suggested.

I've added a junit4 table with the different scenarios so that it has some examples on it. There is a different issue I found (when miss-using the junit4 displayName to not contain the parameter; causing same problem.. it is possible to avoid it, but requires messing with the UniqueID .. probably not the best since it may change... anyways, when attempting 2nd run on the failed tests (even with proper handling), junit4 fails to run only the failed and re-runs all the ones that match the 'displayName' ... but again, since this is some very odd case that I hope nobody is using (And should be discouraged I guess by some warnings when not adding {0} or anything to it)

Will run some more tests and see if I can update this PR to use the newly created branch with less changes or if I should overwrite my branch's history ... will update this PR soon

Thanks!